.github/shared/src/doc-preview.js

```diff
@@ -0,0 +1,138 @@
+  const m = specPath.match(SPEC_FILE_REGEX);
+  if (!m) {
+    console.log(`Path "${specPath}" does not match expected swagger file pattern.`);
+    return null;
+  }
+  const [path, , serviceName, serviceType, resourceProvider, releaseState, apiVersion, , fileName] =
+    m;
+  return {
+    path,
+    serviceName,
+    serviceType,
+    resourceProvider,
+    releaseState,
+    apiVersion,
+    fileName,
+  };
+}
+
+/**
+ * @param {{repoName: string, prNumber: string}} prKey
+ * @returns {object}
+ */
+export function repoJSONTemplate({ repoName, prNumber }) {
+  return {
+    repo: [
+      {
+        url: `https://github.com/${repoName}`,
+        prNumber: prNumber,
+        name: DOCS_NAMESPACE,
+      },
+    ],
+  };
+}
+
+/**
+ * @param {string[]} files
+ * @returns {object}
+ */
+export function mappingJSONTemplate(files) {
+  return {
+    target_api_root_dir: "structured",
+    enable_markdown_fragment: true,
+    markdown_fragment_folder: "authored",
+    use_yaml_toc: true,
+    formalize_url: true,
+    version_list: ["default"],
+    organizations: [
+      {
+        index: "index.md",
+        default_toc_title: "Getting Started",
+        version: "default",
+        services: [
+          {
+            toc_title: "Documentation Preview",
+            url_group: "documentation-preview",
+            swagger_files: files.map((source) => ({
+              source: `${DOCS_NAMESPACE}/${source}`,
+            })),
+          },
+        ],
+      },
+    ],
+  };
+}
+
+/**
+ * @param {string} buildId
+ * @param {{repoName: string, prNumber: string}} key
+ * @returns {string}
+ */
+export function indexMd(buildId, { repoName, prNumber }) {
+  return `# Documentation Preview for swagger pipeline build #${buildId}
+
+Welcome to documentation preview for ${repoName}/pull/${prNumber} 
+created via the swagger pipeline.
+
+Your documentation may be viewed in the menu on the left hand side.
+
+If you have issues around documentation generation, please feel free to contact 
+us in the [Docs Support Teams Channel](https://aka.ms/ci-fix/api-docs-help)`;
+}
+
+/**
+ * Given a list of changed swagger files, select an API version and a list of
+ * swagger files in that API version to process.
+ * @param {string[]} swaggerFiles
+ **/
+export function getSwaggersToProcess(swaggerFiles) {
+  // swaggerFileObjs never has any `null` elements, otherwise, returns the
+  // output type of `parseSwaggerFilePath`.
+  /** @type {Exclude<ReturnType<typeof parseSwaggerFilePath>, null>[]} */
+  const swaggerFileObjs = swaggerFiles.map(parseSwaggerFilePath).filter((obj) => obj !== null);
+
+  const versions = swaggerFileObjs.map((obj) => obj.apiVersion).filter(Boolean);
+  const uniqueVersions = [...new Set(versions)];
+  if (uniqueVersions.length === 0) {
+    console.log(
+      "No API versions found in eligible swagger files. No documentation artifacts will be written.",
+    );
+    return { selectedVersion: null, swaggersToProcess: [] };
```

Comment:
exception and throw, or expected?

Reply:
Getting this far and not throwing could have negative consequences in the docs build. To that end, it's safer to throw and see what situations lead to this throw and dealing with them if they should arise.